utils/src/bitmap/mod.rs

```diff
@@ -0,0 +1,1705 @@
+//! Bitmap implementation
+//!
+//! The bitmap is a compact representation of a sequence of bits, using chunks of bytes for a
+//! more-efficient memory layout than doing [`Vec<bool>`].
+
+use crate::NZUsize;
+#[cfg(not(feature = "std"))]
+use alloc::{collections::VecDeque, vec::Vec};
+use bytes::{Buf, BufMut};
+use commonware_codec::{EncodeSize, Error as CodecError, RangeCfg, Read, ReadExt, Write};
+use core::{
+    fmt::{self, Formatter, Write as _},
+    num::NonZeroUsize,
+    ops::{BitAnd, BitOr, BitXor, Index},
+};
+#[cfg(feature = "std")]
+use std::collections::VecDeque;
+
+mod prunable;
+pub use prunable::Prunable;
+
+/// A bitmap that stores data in chunks of N bytes.
+#[derive(Clone)]
+pub struct BitMap<const N: usize> {
+    /// The bitmap itself, in chunks of size N bytes. The number of valid bits in the last chunk is
+    /// given by `self.next_bit`. Within each byte, lowest order bits are treated as coming before
+    /// higher order bits in the bit ordering.
+    ///
+    /// Invariant: The last chunk in the bitmap always has room for at least one more bit.
+    /// This implies that !chunks.is_empty() always holds.
+    chunks: VecDeque<[u8; N]>,
+
+    /// The position within the last chunk where the next bit is to be appended.
+    ///
+    /// Invariant: This value is always in the range [0, N * 8).
+    next_bit: u64,
+}
+
+impl<const N: usize> BitMap<N> {
+    /// The size of a chunk in bits.
+    pub const CHUNK_SIZE_BITS: u64 = N as u64 * 8;
+
+    /// A chunk of all 0s.
+    const EMPTY_CHUNK: [u8; N] = [0u8; N];
+
+    /// A chunk of all 1s.
+    const FULL_CHUNK: [u8; N] = [u8::MAX; N];
+
+    /// Create a new empty bitmap.
+    pub fn new() -> Self {
+        // Invariant: chunks is never empty
+        let bitmap = VecDeque::from([Self::EMPTY_CHUNK]);
+        Self {
+            chunks: bitmap,
+            next_bit: 0,
+        }
+    }
+
+    // Create a new empty bitmap with the capacity to hold `size` bits without reallocating.
+    pub fn with_capacity(size: u64) -> Self {
```

Comment:
Personally I hate using `u64` everywhere instead of `usize`

Reply:
Changed the API to take in/return `usize`